content/en/blog/_posts/2024-08-20-websocket-transition.md

```diff
@@ -0,0 +1,107 @@
+standardized. Many modern proxies, gateways, and load balancers no longer
+support the protocol. As a result, you might notice that commands like
+`kubectl cp`, `kubectl attach`, `kubectl exec`, and `kubectl port-forward`
+stop working when you try to access your cluster through a proxy or gateway.
+
+As of Kubernetes 1.31, we're modifying the streaming protocol that Kubernetes
+uses for these commands to the more modern [WebSocket streaming protocol](https://datatracker.ietf.org/doc/html/rfc6455).
+The WebSocket protocol is a currently supported standardized streaming protocol
+that guarantees compatibility and interoperability with different components and
+programming languages. The WebSocket protocol is more widely supported by modern
+proxies and gateways than SPDY.
+
+## How streaming APIs work
+
+Kubernetes upgrades HTTP connections to streaming connections by adding
+specific upgrade headers to the originating HTTP request. For example, an HTTP
+upgrade request for running the `date` command on an `nginx` container within
+a cluster is similar to the following:
+
+```shell
+$ kubectl exec -v=8 nginx -- date
+GET https://127.0.0.1:43251/api/v1/namespaces/default/pods/nginx/exec?command=date…
+Request Headers:
+    Connection: Upgrade
+    Upgrade: websocket
+    Sec-Websocket-Protocol: v5.channel.k8s.io
+    User-Agent: kubectl/v1.30.1 (linux/amd64) kubernetes/6911225
+```
+
+If the container runtime supports the WebSocket streaming protocol and at least
+one of the subprotocol versions (e.g. `v5.channel.k8s.io`), the server responds
+with a successful `101 Switching Protocols` status, along with the negotiated
+subprotocol version:
+
+```shell
+Response Status: 101 Switching Protocols in 3 milliseconds
+Response Headers:
+    Upgrade: websocket
+    Connection: Upgrade
+    Sec-Websocket-Accept: j0/jHW9RpaUoGsUAv97EcKw8jFM=
+    Sec-Websocket-Protocol: v5.channel.k8s.io
+```
+
+At this point the TCP connection used for the HTTP protocol has changed to a
+streaming connection. Subsequent STDIN, STDOUT, and STDERR data (as well as TTY
+and process exit code data) for this shell interaction is then streamed over this
+upgraded connection.
+
+## How to use the new WebSocket streaming protocol
+
+If your cluster and kubectl are on version 1.29 or later, there are two
+control plane feature flags and two kubectl feature flags that govern
+the use of the new WebSocket streaming protocol. In version 1.31 and
+later, all of the following feature flags are in beta and are enabled by
+default:
+
+- Control Plane Feature Flags
+  - TranslateStreamCloseWebsocketRequests
+      - `.../exec`
+      - `.../attach`
+  - PortForwardWebsockets
+      - `.../port-forward`
+- kubectl Feature Flags (environment variables)
+  - `KUBECTL_REMOTE_COMMAND_WEBSOCKETS`
+      - `kubectl exec`
+      - `kubectl cp`
+      - `kubectl attach`
+  - `KUBECTL_PORT_FORWARD_WEBSOCKETS`
+      - `kubectl port-forward`
+
```

Comment:
(optionally)
```suggestion

If you're connecting to an older cluster but can manage the feature gate settings, turn on
`TranslateStreamCloseWebsocketRequests` and `PortForwardWebsockets` to try this new
behavior (you'll need Kubernetes v1.29 or v1.30 to try it out).

```

Reply:
I added this small paragraph, but there is a subtle distinction I added. The `TranslateStreamCloseWebsocketRequests` feature gate is the only one available in v1.29. The `PortForwardWebsockets` feature gate is not available until v1.30. Please have a look at what I wrote, and let me know what you think.